server/src/main/java/org/elasticsearch/action/bulk/Retry2.java

```diff
@@ -0,0 +1,297 @@
+     * either all requests have succeeded or all retry attempts have been exhausted.
+     */
+    private final class RetryHandler implements ActionListener<BulkResponse> {
+        private static final RestStatus RETRY_STATUS = RestStatus.TOO_MANY_REQUESTS;
+        private final BulkRequest bulkRequest;
+        private final BiConsumer<BulkRequest, ActionListener<BulkResponse>> consumer;
+        private final ActionListener<BulkResponse> listener;
+        private final List<BulkItemResponse> responsesAccumulator;
+        private final long startTimestampNanos;
+        private final int retriesRemaining;
+
+        /**
+         * Creates a RetryHandler listener
+         * @param bulkRequest The BulkRequest to be sent, a subset of the original BulkRequest.
+         * @param responsesAccumulator The accumulator of all BulkItemResponses for the original BulkRequest
+         * @param consumer
+         * @param listener The delegate listener
+         * @param retriesRemaining The number of retry attempts remaining for the bulkRequestForRetry
+         */
+        RetryHandler(
+            BulkRequest bulkRequest,
+            List<BulkItemResponse> responsesAccumulator,
+            BiConsumer<BulkRequest, ActionListener<BulkResponse>> consumer,
+            ActionListener<BulkResponse> listener,
+            int retriesRemaining
+        ) {
+            this.bulkRequest = bulkRequest;
+            this.responsesAccumulator = responsesAccumulator;
+            this.consumer = consumer;
+            this.listener = listener;
+            this.startTimestampNanos = System.nanoTime();
+            this.retriesRemaining = retriesRemaining;
+        }
+
+        @Override
+        public void onResponse(BulkResponse bulkItemResponses) {
+            if (bulkItemResponses.hasFailures() == false) {
+                totalBytesInFlight.addAndGet(-1 * bulkRequest.estimatedSizeInBytes());
+                logger.trace(
+                    "Got a response in {} with {} items, no failures",
+                    bulkItemResponses.getTook(),
+                    bulkItemResponses.getItems().length
+                );
+                // we're done here, include all responses
+                addResponses(bulkItemResponses, (r -> true));
+                listener.onResponse(getAccumulatedResponse());
+            } else {
+                if (canRetry(bulkItemResponses)) {
+                    logger.trace(
+                        "Got a response in {} with {} items including failures, can retry",
+                        bulkItemResponses.getTook(),
+                        bulkItemResponses.getItems().length
+                    );
+                    addResponses(bulkItemResponses, (r -> r.isFailed() == false));
+                    retry(createBulkRequestForRetry(bulkItemResponses), responsesAccumulator, consumer, listener, retriesRemaining);
```

Comment:
This looks like it leaks the bytes corresponding with the successful items, because `createBulkRequestForRetry(bulkItemResponses)` will be smaller than the original request.

Reply:
I don't think I'm following -- leaking in what way? We're intentionally putting them in the responsesAccumulator so that we can notify the listener once for the whole bulk once we know the final result. Only the failed ones are getting added to the request to retry right?

Reply:
AFAICT on each retry we replace the bulk request with a _possibly-smaller_ bulk request, namely `createBulkRequestForRetry(bulkItemResponses)`. If such an attempt finally completes, we only release the `estimatedSizeInBytes()` for the very last bulk request. But we originally acquired the `estimatedSizeInBytes()` for the first attempt, which is in general going to be larger.